hip_data_tools/google/adwords.py

```diff
@@ -579,3 +583,69 @@ def set_query_to_fetch_all(self, start_index: int = 0, page_size: int = 500):
                  .Limit(start_index, page_size)
                  .Build())
         self.set_query(query)
+
+
+class AdWordsReportDefinitionReader(AdWordsUtil):
+    """
+    Class to interact with the ReportDefinitionService and access fields for adwords reports
+    Args:
+        conn (GoogleAdWordsConnectionManager): Connection manager to handle the creation of
+    adwords client
+    """
+
+    def __init__(self, conn: GoogleAdWordsConnectionManager):
+        super().__init__(conn, "ReportDefinitionService", "v201809")
+
+    def get_report_fields(self, report_type: str) -> List[dict]:
+        """
+        Get a list of fields for a given report.
+        Args:
+            report_type (str): Possible reports are defined in the documentation here -
+            https://developers.google.com/adwords/api/docs/appendix/reports/all-reports
+        Returns: List[dict]
+        """
+        return self._get_service().getReportFields(report_type)
+
+
+class AdWordsReportReader:
+    """
+    Generic data reader class for downloading data from report awql efficiently
+    Args:
+            conn (GoogleAdWordsConnectionManager): Connection manager to handle the creation of
+        adwords client
+    """
+
+    def __init__(self, conn: GoogleAdWordsConnectionManager):
+        self.conn = conn
+        self.version = 'v201809'
+        self.__downloader = None
+
+    def _get_report_downloader(self):
+        if self.__downloader is None:
+            client = self.conn.get_adwords_client()
+            self.__downloader = client.GetReportDownloader(version=self.version)
+        return self.__downloader
+
+    def awql_to_dataframe(self, query: ReportQuery) -> DataFrame:
+        """
+        Download the data returned by report query in a compressed format and return it in form of
+        a pandas dataframe
+        Args:
+            query (ReportQuery): an awql report query, see example
+            https://github.com/googleads/googleads-python-lib/blob/master/examples/adwords
+            /v201809/reporting/stream_criteria_report_results.py#L34-L39
+        Returns: DataFrame
+        """
+        with NamedTemporaryFile(mode="w+b") as temp_file:
+            self._get_report_downloader().DownloadReportWithAwql(
+                query,
+                "GZIPPED_CSV",
+                temp_file,
+                skip_report_header=False,
+                skip_column_header=False,
+                skip_report_summary=False,
+                include_zero_impressions=True
```

Comment:
are there reports where this would be an invalid option?

Reply:
seems to work on all reports, even the ones which don't have impressions. I think its ignored if not required.